src/box/replication.cc

```diff
@@ -1362,18 +1353,13 @@ void
 replica_on_relay_stop(struct replica *replica)
 {
 	/*
-	 * If the replica was evicted from the cluster, or was not
-	 * even added there (anon replica), we don't need to keep
-	 * WALs for it anymore. Unregister it with the garbage
-	 * collector then. See also replica_clear_id.
+	 * If the replica was evicted from the cluster, we don't need to keep
+	 * WALs for it anymore. Unregister it with the garbage collector then.
+	 * See also replica_clear_id.
 	 */
-	if (replica->id == REPLICA_ID_NIL) {
-		if (!replica->anon) {
-			gc_consumer_unregister(replica->gc);
-			replica->gc = NULL;
-		} else {
-			assert(replica->gc == NULL);
-		}
+	if (!replica->anon && replica->id == REPLICA_ID_NIL) {
+		gc_consumer_unregister(replica->gc);
+		replica->gc = NULL;
 	}
```

Comment:
> AFAIU, this function is supposed to be called when the replica is evicted from the cluster.

The function is called not only when a replica is evicted from the cluster - it is called on relay stop, IOW after each subscribe. So, if we remove GC consumer of anonymous replicas here, they will be pointless since being deleted on every disconnect.

> The fact that an anon replica being evicted from the cluster still retains the consumer

Anon replica does not participate in the cluster by definition, hence, cannot be evicted.

What actually this diff does is simply removing an assertion that anonymous replicas have no GC consumers. For non-anonymous replicas, the logic remains - we remove GC consumer in [`replica_clear_id`](https://github.com/tarantool/tarantool/blob/f69e2ae488b3620e31f1a599d8fb78a66917dbfd/src/box/replication.cc#L451) only if the replica's relay is disconnected, otherwise it is deleted here. Condition `!replica->anon && replica->id == REPLICA_ID_NIL` basically means "the replica was evicted from the cluster".